src/Compiler/Microsoft.AspNetCore.Mvc.Razor.Extensions.Version1_X/src/InjectDirective.cs

```diff
@@ -71,7 +71,14 @@ protected override void ExecuteCore(RazorCodeDocument codeDocument, DocumentInte
                     continue;
                 }
 
-                typeName = typeName.Replace("<TModel>", "<" + modelType + ">");
+                if (typeName.EndsWith("<TModel>", StringComparison.Ordinal))
+                {
+                    typeName = typeName.Replace("<TModel>", "<" + modelType + ">");
+                    if (typeSpan.HasValue)
+                    {
+                        typeSpan = new SourceSpan(typeSpan.Value.FilePath, typeSpan.Value.AbsoluteIndex, typeSpan.Value.LineIndex, typeSpan.Value.CharacterIndex, typeSpan.Value.Length - 8, typeSpan.Value.LineCount, typeSpan.Value.EndCharacterIndex - 8);
+                    }
+                }
```

Comment:
Nit: consider using a constant for `"<TModel>"` and its Length; and replace just the trailing `<TModel>`, not potentially all `<TModel>` substrings in `typeName`. Something like this (haven't tested the code, might have some off by one errors):

```suggestion
                const string tModel = "<TModel>";
                if (typeName.EndsWith(tModel, StringComparison.Ordinal))
                {
                    typeName = typeName[..^(tModel.Length)] + "<" + modelType + ">";
                    if (typeSpan.HasValue)
                    {
                        typeSpan = new SourceSpan(typeSpan.Value.FilePath, typeSpan.Value.AbsoluteIndex, typeSpan.Value.LineIndex, typeSpan.Value.CharacterIndex, typeSpan.Value.Length - tModel.Length, typeSpan.Value.LineCount, typeSpan.Value.EndCharacterIndex - tModel.Length);
                    }
                }
```